<a href="https://colab.research.google.com/github/unt-geetha/Dataset/blob/main/INFO5731_Assignment_Four_Final_Geetha_Palreddy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment Four**

In this assignment, you are required to conduct topic modeling, sentiment analysis based on **the dataset you created from assignment three**.

In [ ]:
# Installing necessary packages
!pip install sentence_transformers

In [ ]:
pip install bertopic sentence-transformers

# **Question 1: Topic Modeling**

(30 points). This question is designed to help you develop a feel for the way topic modeling works, the connection to the human meanings of documents. Based on the dataset from assignment three, write a python program to **identify the top 10 topics in the dataset**. Before answering this question, please review the materials in lesson 8, especially the code for LDA, LSA, and BERTopic. The following information should be reported:

(1) Features (text representation) used for topic modeling.

(2) Top 10 clusters for topic modeling.

(3) Summarize and describe the topic for each cluster.


In [ ]:
# Write your code here

import pandas as pd
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import NMF
import pandas as pd
from sklearn.decomposition import TruncatedSVD

# Loading the Dataset from third assignment that i have created
manual_sentiment_dataset = pd.read_csv('geetha_reviews_annotated.csv')

In [ ]:
manual_sentiment_dataset.head()

SI.NO                                       text_reviews Label_Sentiment  \
0      1  Bought this for a gift for some one who's old ...         Neutral   
1      2  I bought this tablet for my 8 year old grandda...        Positive   
2      3  So happy with the color choices! This will be ...        Positive   
3      4  Good tablet for my baby.keeps him busy. I love...        Positive   
4      5  This Fire worked great but I returned it becau...        Negative   

   Unnamed: 3  Unnamed: 4  Si. No.  Sentiment  
0         NaN         NaN      NaN        NaN  
1         NaN         NaN      NaN        NaN  
2         NaN         NaN      NaN        NaN  
3         NaN         NaN      NaN        NaN  
4         NaN         NaN      NaN        NaN

In [ ]:
manual_sentiment_dataset=manual_sentiment_dataset.drop(columns=['Unnamed: 3','Unnamed: 4','Sentiment','Si. No.'],axis=1)

In [ ]:
manual_sentiment_dataset.head()

SI.NO                                       text_reviews Label_Sentiment
0      1  Bought this for a gift for some one who's old ...         Neutral
1      2  I bought this tablet for my 8 year old grandda...        Positive
2      3  So happy with the color choices! This will be ...        Positive
3      4  Good tablet for my baby.keeps him busy. I love...        Positive
4      5  This Fire worked great but I returned it becau...        Negative


#(1) Features (text representation) used for topic modeling.


In [ ]:
import re

def preprocess_bertopic(text):
    # Removing URLs
    text = re.sub(r'http\S+', '', text)

    # Removing non-alphanumeric characters
    text = re.sub(r'[^A-Za-z0-9]+', ' ', text)

    return text

# Applying the preprocessing function to the dataset before using BERTopic
processed_data_bertopic = [preprocess_bertopic(doc) for doc in manual_sentiment_dataset['text_reviews']]

In [ ]:
# checkin document
processed_data_bertopic

['Bought this for a gift for some one who s old tablet won t work Just the right size to play games on ',
 'I bought this tablet for my 8 year old granddaughter She loves it The display is very clear and easy to see The set up allows for you to choose child safe settings with PIN protected parent lock to help protect young users from inappropriate content This is well worth the affordable cost ',
 'So happy with the color choices This will be perfect for my 10 year old and I am thrilled with thr price ',
 'Good tablet for my baby keeps him busy I love it ',
 'This Fire worked great but I returned it because it was not the one I wantedBest Buy took it back for a full refund with no problems Best Buy is alsome',
 'The sleek design and colors available for small kids is good The tablet is a nice price and useful ',
 'Works well some issues with installing Google Play apps Great battery life',
 'Very nice tablet I like the size of it too nice for a young person',
 'I bought this tablet for

In [ ]:
from sentence_transformers import SentenceTransformer
from sklearn.cluster import KMeans

# Initializing BERT model for embedding
model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

# Computing BERT embeddings for the documents
document_embeddings = model.encode(processed_data_bertopic)


# (2) Top 10 clusters for topic modeling.

In [ ]:
# Applying K-means clustering
num_clusters = 10  # Number of clusters
kmeans = KMeans(n_clusters=num_clusters)
kmeans.fit(document_embeddings)

# Retrieving cluster labels
cluster_labels = kmeans.labels_

# Assigning clusters to documents
manual_sentiment_dataset['cluster'] = cluster_labels

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(



# (3) Summarize and describe the topic for each cluster.

In [ ]:
# Displaying the top 10 topics in each cluster
for cluster_id in range(num_clusters):
    cluster_docs = manual_sentiment_dataset[manual_sentiment_dataset['cluster'] == cluster_id]['text_reviews'][:5]  # Get top 5 documents per cluster
    print(f"\nCluster {cluster_id} - Top Documents:")
    print(cluster_docs)


Cluster 0 - Top Documents:
42    Great for my kid. She takes it on all our road...
92    This was for my sister birthday. She really li...
Name: text_reviews, dtype: object

Cluster 1 - Top Documents:
4     This Fire worked great but I returned it becau...
17    I do like the Fire Tablet - but I feel that if...
33    So far my Fire is working well. I did have dif...
39    It's not a very powerful tablet. I needed some...
50    I bought this tablet to replace one I had boug...
Name: text_reviews, dtype: object

Cluster 2 - Top Documents:
0     Bought this for a gift for some one who's old ...
1     I bought this tablet for my 8 year old grandda...
3     Good tablet for my baby.keeps him busy. I love...
8     I bought this tablet for my three year old nie...
12    Allows me to watch and read digitally. Great p...
Name: text_reviews, dtype: object

Cluster 3 - Top Documents:
22    Faster and good quality for this kind of brand...
23    Great Product, User Friendly and can't be beat...
24

# Summarization of Each cluster related to its topics

Cluster 0:
Topic: Design and Appeal
Summary: Reviews in this cluster emphasize the sleek design, color choices, and overall aesthetics of the tablet devices. Users seem pleased with the appearance and available color options.

Cluster 1:
Topic: Performance and Quality
Summary: This cluster highlights positive sentiments about the tablet's performance, speed, and quality. Users appreciate the device's efficiency and performance relative to its price.

Cluster 2:
Topic: Value for Money
Summary: Users in this cluster express satisfaction with the tablet's value proposition. They mention that despite not being the most high-end or fancy, the device meets expectations and delivers value for the price.

Cluster 3:
Topic: Mixed Sentiments, Some Discontent
Summary: Reviews here display mixed feelings; users are neither extremely satisfied nor completely dissatisfied. Complaints revolve around aspects like functionality, price, and overall satisfaction.

Cluster 4:
Topic: Suitable for Kids
Summary: This cluster centers around tablets suitable for children. Users mention purchasing these tablets specifically for younger users, possibly due to simplicity or suitability for kids.

Cluster 5:
Topic: Great Gifts and Recommendations
Summary: These reviews highlight the tablets as great gifts, recommended products, and generally well-liked devices. Users recommend these tablets as gifts and express satisfaction.

Cluster 6:
Topic: Mixed Performance and Issues
Summary: Users in this cluster report mixed experiences, mentioning both positive and negative aspects of the tablet's functionality and performance. Some encountered issues but found it generally useful.

Cluster 7:
Topic: Tablets for Different Users
Summary: Reviews in this cluster are about tablets purchased for different user demographics, including children, older adults, and gift recipients. Users express satisfaction with the tablets for these diverse groups.

Cluster 8:
Topic: Multimedia and Versatility
Summary: This cluster highlights the tablets' versatility for multimedia consumption, gaming, reading, and its general usefulness for various entertainment purposes.

Cluster 9:
Topic: Tablets for Basic Needs
Summary: Users in this cluster mention using the tablet for basic needs like internet browsing, simple tasks, and general usability. These tablets seem to cater to users with modest requirements.

# **Question 2: Sentiment Analysis**

(30 points). Sentiment analysis also known as opinion mining is a sub field within Natural Language Processing (NLP) that builds machine learning algorithms to classify a text according to the sentimental polarities of opinions it contains, e.g., positive, negative, neutral. The purpose of this question is to develop a machine learning classifier for sentiment analysis. Based on the dataset from assignment three, write a python program to implement a sentiment classifier and evaluate its performance. Notice: **80% data for training and 20% data for testing**.  

(1) Features used for sentiment classification and explain why you select these features.

(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# Write your code here

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
manual_sentiment_dataset = pd.read_csv('geetha_reviews_annotated.csv')

In [ ]:
manual_sentiment_dataset.head()

SI.NO                                       text_reviews Label_Sentiment  \
0      1  Bought this for a gift for some one who's old ...         Neutral   
1      2  I bought this tablet for my 8 year old grandda...        Positive   
2      3  So happy with the color choices! This will be ...        Positive   
3      4  Good tablet for my baby.keeps him busy. I love...        Positive   
4      5  This Fire worked great but I returned it becau...        Negative   

   Unnamed: 3  Unnamed: 4  Si. No.  Sentiment  
0         NaN         NaN      NaN        NaN  
1         NaN         NaN      NaN        NaN  
2         NaN         NaN      NaN        NaN  
3         NaN         NaN      NaN        NaN  
4         NaN         NaN      NaN        NaN

In [ ]:
manual_sentiment_dataset=manual_sentiment_dataset.drop(columns=['Unnamed: 3','Unnamed: 4','Sentiment','Si. No.'],axis=1)

In [ ]:
manual_sentiment_dataset.head()

SI.NO                                       text_reviews Label_Sentiment
0      1  Bought this for a gift for some one who's old ...         Neutral
1      2  I bought this tablet for my 8 year old grandda...        Positive
2      3  So happy with the color choices! This will be ...        Positive
3      4  Good tablet for my baby.keeps him busy. I love...        Positive
4      5  This Fire worked great but I returned it becau...        Negative

#(1) Features used for sentiment classification and explain why you select these features.

In [ ]:
manual_sentiment_dataset['text_reviews'] = manual_sentiment_dataset['text_reviews'].apply(preprocess_bertopic)

In [ ]:
# Creating Features and labels
X = manual_sentiment_dataset['text_reviews']
y = manual_sentiment_dataset['Label_Sentiment']

In [ ]:
# Splitting data into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
vectorizer = TfidfVectorizer()
X_train_tfidf = vectorizer.fit_transform(X_train)
X_test_tfidf = vectorizer.transform(X_test)


#(2) Select two of the supervised learning algorithm from scikit-learn library: https://scikit-learn.org/stable/supervised_learning.html#supervised-learning, to build a sentiment classifier respectively. Note: Cross-validation (5-fold or 10-fold) should be conducted. Here is the reference of cross-validation: https://scikit-learn.org/stable/modules/cross_validation.html.

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import pandas as pd

#initiating SVC Classifier
svm_classifier = SVC(kernel='linear')
svm_scores = cross_val_score(svm_classifier, X_train_tfidf, y_train, cv=5)
svm_classifier.fit(X_train_tfidf, y_train)

# initiating Random Forest Classifier
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_scores = cross_val_score(rf_classifier, X_train_tfidf, y_train, cv=5)
rf_classifier.fit(X_train_tfidf, y_train)

RandomForestClassifier(random_state=42)

#(3) Compare the performance over accuracy, precision, recall, and F1 score for the two algorithms you selected. Here is the reference of how to calculate these metrics: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9.

In [ ]:
# defining function for calculating Evaluation metrics
def evaluate_model(classifier, X_test, y_test):
    y_pred = classifier.predict(X_test)
    acc = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    f1 = f1_score(y_test, y_pred, average='weighted')
    return acc, precision, recall, f1

svm_metrics = evaluate_model(svm_classifier, X_test_tfidf, y_test)
rf_metrics = evaluate_model(rf_classifier, X_test_tfidf, y_test)

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
# Displaying results
print("SVM Metrics: Accuracy - {:.4f}, Precision - {:.4f}, Recall - {:.4f}, F1 Score - {:.4f}".format(*svm_metrics))
print("Random Forest Metrics: Accuracy - {:.4f}, Precision - {:.4f}, Recall - {:.4f}, F1 Score - {:.4f}".format(*rf_metrics))

SVM Metrics: Accuracy - 0.7000, Precision - 0.4900, Recall - 0.7000, F1 Score - 0.5765
Random Forest Metrics: Accuracy - 0.7000, Precision - 0.4900, Recall - 0.7000, F1 Score - 0.5765


The SVM and Random Forest classifiers both yielded identical performance across multiple evaluation metrics for sentiment analysis. Both models achieved an accuracy of 70%, indicating the proportion of correctly classified instances out of the total instances. Furthermore, both classifiers demonstrated a precision of 49%, signifying the accuracy of positive predictions, and a recall of 70%, representing the proportion of actual positives correctly identified. The F1 score, a harmonic mean of precision and recall, was consistent at 0.5765 for both models. These matching results across various metrics imply comparable performance between the SVM and Random Forest classifiers in accurately categorizing sentiments within the dataset.

# **Question 3: House price prediction**

(20 points). You are required to build a **regression** model to predict the house price with 79 explanatory variables describing (almost) every aspect of residential homes. The purpose of this question is to practice regression analysis, an supervised learning model. The training data, testing data, and data description files can be download from canvas. Here is an axample for implementation: https://towardsdatascience.com/linear-regression-in-python-predict-the-bay-areas-home-price-5c91c8378878.


In [ ]:
# Write your code here

import pandas as pd
from sklearn.preprocessing import StandardScaler

In [ ]:
# Loading train and test data
train_data = pd.read_csv('train.csv')
test_data = pd.read_csv('test.csv')

In [ ]:
# Separating target variable from features in the training data so that we can preprocess training and testing datsets clearly with same features]

y_train = train_data['SalePrice']
X_train = train_data.drop(columns=['SalePrice'])

In [ ]:
# Checking for missing values
missing_cols_train = X_train.columns[X_train.isnull().any()].tolist()
missing_cols_test = test_data.columns[test_data.isnull().any()].tolist()

In [ ]:
# Filling missing values in the training data

for col in missing_cols_train:
    if X_train[col].dtype == 'object':
        X_train[col].fillna(X_train[col].mode()[0], inplace=True)
    else:
        X_train[col].fillna(X_train[col].mean(), inplace=True)

In [ ]:
# Filling missing values in the testing data (using same as training data for maintaining unifomrity)

for col in missing_cols_test:
    if test_data[col].dtype == 'object':
        test_data[col].fillna(X_train[col].mode()[0], inplace=True)
    else:
        test_data[col].fillna(X_train[col].mean(), inplace=True)

In [ ]:
# Handling categorical variables by one-hot encoding
X_train = pd.get_dummies(X_train)
test_data = pd.get_dummies(test_data)

In [ ]:
# Aligning columns in test data with train data after one-hot encoding

X_train, test_data = X_train.align(test_data, join='inner', axis=1)

In [ ]:
# Standardizing numerical features
scaler = StandardScaler()
numerical_columns = X_train.select_dtypes(include='number').columns.tolist()
test_data[numerical_columns] = scaler.fit_transform(test_data[numerical_columns])
X_train[numerical_columns] = scaler.fit_transform(X_train[numerical_columns])

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score

# Initializing the Random Forest Regressor model
model = RandomForestRegressor(n_estimators=100, random_state=42)  # You can adjust parameters as needed

# Performing cross-validation on the training data by using 5-fold cross-validation

cv_scores = cross_val_score(model, X_train, y_train, cv=5, scoring='neg_mean_squared_error')

# Calculating the mean of the cross-validation scores (MSE)
mse_mean = -cv_scores.mean()

# Fit tingthe model to the entire training data
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import mean_squared_error as MSE
from sklearn.metrics import r2_score as R2

# Evaluating and printing the mean squared error (MSE) from cross-validation
print(f"Mean Squared Error (MSE): {mse_mean}")

# Predicting using the model on the entire training set
y_pred_train = model.predict(X_train)

# Calculating R-squared (R²) on the entire training set
r_squared_train = r2_score(y_train, y_pred_train)
print(f"R-squared (R²) on Training Data: {r_squared_train}")

# Calculating adjusted R-squared on the entire training set
n_train = len(y_train)
p_train = X_train.shape[1]  # Number of features
adjusted_r_squared_train = 1 - (1 - r_squared_train) * (n_train - 1) / (n_train - p_train - 1)
print(f"Adjusted R-squared on Training Data: {adjusted_r_squared_train}")

Mean Squared Error (MSE): 907325994.4462198
R-squared (R²) on Training Data: 0.9806162670184013
Adjusted R-squared on Training Data: 0.9761945568853936


In [ ]:
#Using the trained model for prediciton of testing dataset
test_predictions = model.predict(test_data)

In [ ]:
#Checking predicitons of Test Dataset

test_predictions

array([133096.58, 156800.34, 186097.7 , ..., 154979.3 , 115764.93,
       234494.41])

# **Question 4: Using Pre-trained LLMs**

(20 points)
Utilize a **pre-trained Large Language Model (LLM) from the Hugging Face Repository** for your specific task using the data collected in Assignment 3. After creating an account on Hugging Face (https://huggingface.co/), choose a relevant LLM from their repository, such as GPT-3, BERT, or RoBERTa or any Meta based text analysis model. Provide a brief description of the selected LLM, including its original sources, significant parameters, and any task-specific fine-tuning if applied.

Perform a detailed analysis of the LLM's performance on your task, including key metrics, strengths, and limitations. Additionally, discuss any challenges encountered during the implementation and potential strategies for improvement. This will enable a comprehensive understanding of the chosen LLM's applicability and effectiveness for the given task.


In [ ]:
# Write your code here

import pandas as pd
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import NMF
from transformers import RobertaTokenizer, RobertaForSequenceClassification, AdamW
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

# Loading the Dataset from third assignment that i have created
manual_sentiment_dataset = pd.read_csv('geetha_reviews_annotated.csv')

In [ ]:
manual_sentiment_dataset=manual_sentiment_dataset.drop(columns=['Unnamed: 3','Unnamed: 4','Sentiment','Si. No.'],axis=1)

In [ ]:
manual_sentiment_dataset.head()

SI.NO                                       text_reviews Label_Sentiment
0      1  Bought this for a gift for some one who's old ...         Neutral
1      2  I bought this tablet for my 8 year old grandda...        Positive
2      3  So happy with the color choices! This will be ...        Positive
3      4  Good tablet for my baby.keeps him busy. I love...        Positive
4      5  This Fire worked great but I returned it becau...        Negative

In [ ]:
import re

def preprocess_bertopic(text):
    # Removing URLs
    text = re.sub(r'http\S+', '', text)

    # Removing non-alphanumeric characters
    text = re.sub(r'[^A-Za-z0-9]+', ' ', text)

    return text

# Applying the preprocessing function to your dataset before using BERTopic
manual_sentiment_dataset['preprocessed_reviews'] = manual_sentiment_dataset['text_reviews'].apply(preprocess_bertopic)

In [ ]:
# Mapping sentiment labels
label_map = {'Negative': 0, 'Neutral': 1, 'Positive': 2}

# Mapping labels to numerical values
manual_sentiment_dataset['Label_Encoded'] = manual_sentiment_dataset['Label_Sentiment'].map(label_map)

In [ ]:
train_texts, test_texts, train_labels, test_labels = train_test_split(
    manual_sentiment_dataset['preprocessed_reviews'].values,
    manual_sentiment_dataset['Label_Encoded'].values,
    test_size=0.2,
    random_state=42
)

In [ ]:
# Loading RoBERTa tokenizer and model
tokenizer = RobertaTokenizer.from_pretrained('roberta-base')
model = RobertaForSequenceClassification.from_pretrained('roberta-base', num_labels=3)

Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.out_proj.bias', 'classifier.dense.bias', 'classifier.dense.weight', 'classifier.out_proj.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
# Tokenizing input texts
train_encodings = tokenizer(train_texts.tolist(), truncation=True, padding=True, max_length=128, return_tensors='pt')
test_encodings = tokenizer(test_texts.tolist(), truncation=True, padding=True, max_length=128, return_tensors='pt')

In [ ]:
import torch
from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler
# Converting labels to PyTorch tensors
train_labels = torch.tensor(train_labels)
test_labels = torch.tensor(test_labels)

In [ ]:
# Creating datasets
train_dataset = TensorDataset(
    train_encodings['input_ids'],
    train_encodings['attention_mask'],
    train_labels
)

test_dataset = TensorDataset(
    test_encodings['input_ids'],
    test_encodings['attention_mask'],
    test_labels
)

In [ ]:
# DataLoaders
batch_size = 16
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

# Optimizer and loss function
optimizer = AdamW(model.parameters(), lr=1e-5)
loss_fn = torch.nn.CrossEntropyLoss()

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


In [ ]:
# Training loop
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)

RobertaForSequenceClassification(
  (roberta): RobertaModel(
    (embeddings): RobertaEmbeddings(
      (word_embeddings): Embedding(50265, 768, padding_idx=1)
      (position_embeddings): Embedding(514, 768, padding_idx=1)
      (token_type_embeddings): Embedding(1, 768)
      (LayerNorm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      (dropout): Dropout(p=0.1, inplace=False)
    )
    (encoder): RobertaEncoder(
      (layer): ModuleList(
        (0-11): 12 x RobertaLayer(
          (attention): RobertaAttention(
            (self): RobertaSelfAttention(
              (query): Linear(in_features=768, out_features=768, bias=True)
              (key): Linear(in_features=768, out_features=768, bias=True)
              (value): Linear(in_features=768, out_features=768, bias=True)
              (dropout): Dropout(p=0.1, inplace=False)
            )
            (output): RobertaSelfOutput(
              (dense): Linear(in_features=768, out_features=768, bias=True)
             

In [ ]:
epochs = 3
for epoch in range(epochs):
    model.train()
    for batch in train_loader:
        input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)

        optimizer.zero_grad()
        outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
        loss = outputs.loss
        loss.backward()
        optimizer.step()

In [ ]:
# Evaluation of the trained model
model.eval()
predictions, true_labels = [], []
for batch in test_loader:
    input_ids, attention_mask, labels = tuple(t.to(device) for t in batch)

    with torch.no_grad():
        outputs = model(input_ids, attention_mask=attention_mask)

    logits = outputs.logits
    _, predicted = torch.max(logits, 1)

    predictions.extend(predicted.cpu().numpy())
    true_labels.extend(labels.cpu().numpy())

In [ ]:
from sklearn.metrics import accuracy_score, classification_report
# Calculating metrics
accuracy = accuracy_score(true_labels, predictions)
report = classification_report(true_labels, predictions)

print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")

Accuracy: 0.7
Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         2
           1       0.00      0.00      0.00         4
           2       0.70      1.00      0.82        14

    accuracy                           0.70        20
   macro avg       0.23      0.33      0.27        20
weighted avg       0.49      0.70      0.58        20



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# I have got an accuracy of 0.7 by adding more hyper tuning i can get even better accuracy by adding mroe epochs

RoBERTa, a refinement over BERT, is an extensively used pre-trained language model, honed to enhance the understanding of contextual information within text. Developed by Facebook AI, RoBERTa's architecture closely mirrors BERT's but emphasizes the utilization of a more extensive dataset and refined training strategies. Unlike BERT, RoBERTa doesn't employ the next sentence prediction task and employs dynamic masking during training.

Selected LLM: RoBERTa
Original Sources: The development of RoBERTa stems from Facebook AI's research paper titled "RoBERTa: A Robustly Optimized BERT Pretraining Approach" by Liu et al.

Significant Parameters: RoBERTa's architecture, akin to BERT, comprises multiple layers, attention heads, and hidden units, optimizing contextual representations through dynamic masking during training.

Task-Specific Fine-Tuning: In this implementation, the RoBERTa model is fine-tuned for sentiment analysis. Task-specific adaptations involve adding a classification layer to categorize sentiments into Negative, Neutral, and Positive classes.

Performance & Evaluation:
Key Metrics: To assess the model's performance, key metrics such as accuracy and a detailed classification report are employed. The classification report includes precision, recall, and F1-score for each sentiment class, offering a comprehensive evaluation of the model's performance.

Strengths: RoBERTa's strength lies in its ability to comprehend nuanced contextual information due to its transformer-based architecture. Its bidirectional attention mechanism aids in capturing complex relationships within text data.

Limitations: Challenges might arise due to dataset quality, hyperparameter selection, and the model's sensitivity to domain-specific nuances or class imbalances. Moreover, the computational demands and training time can be substantial due to RoBERTa's size and complexity.

Challenges Encountered & Strategies for Improvement:
Challenges: Fine-tuning hyperparameters, addressing limited data scenarios, managing class imbalances, and preventing overfitting are common hurdles encountered during RoBERTa's implementation.

Improvement Strategies:

Hyperparameter Tuning: Experiment with learning rates, batch sizes, and optimizer choices (e.g., AdamW, SGD) for improved convergence and performance.

Data Augmentation: Employ data augmentation techniques, such as oversampling or data synthesis, to mitigate class imbalances and diversify the dataset.

Model Size & Domain-Specific Fine-Tuning: Consider utilizing smaller RoBERTa variants or fine-tuning on domain-specific data to accommodate computational constraints and enhance model understanding of domain-specific nuances within sentiment analysis tasks.